File: Content Based RS/Function.py
import pandas as pd
import numpy as np
import math

Ratings=pd.read_csv("C:/Users/Flora Huang/Documents/GitHub/Recommender System/Content Based Recommender/Data/ratings.csv")
Movies=pd.read_csv("C:/Users/Flora Huang/Documents/GitHub/Recommender System/Content Based Recommender/Data/movies.csv", encoding = "ISO-8859-1")
Tags=pd.read_csv("C:/Users/Flora Huang/Documents/GitHub/Recommender System/Content Based Recommender/Data/tags.csv", encoding = "ISO-8859-1")

#Calculating the TF value and IDF value and multiplying together to get TF-IDF value
TF= Tags.groupby(["movieId","tag"], as_index = False, sort = False).count().rename(columns = {"userId": "tag_count_TF"})[["movieId","tag","tag_count_TF"]]
Tag_distinct = Tags[["tag","movieId"]].drop_duplicates()
DF =Tag_distinct.groupby(["tag"], as_index = False, sort = False).count().rename(columns = {"movieId": "tag_count_DF"})[["tag","tag_count_DF"]]
a=math.log10(len(np.unique(Tags["movieId"])))
DF["IDF"] = a - np.log10(DF["tag_count_DF"])
TF = pd.merge(TF,DF,on = "tag", how = "left", sort = False)
TF["TF-IDF"]=TF["tag_count_TF"]*TF["IDF"]

#Calculating the unit length vector by dividing TF-IDF value with the vector length of a particular movie.
Vect_len=TF[["movieId","TF-IDF"]].copy()
Vect_len["TF-IDF-Sq"] = Vect_len["TF-IDF"]**2
Vect_len = Vect_len.groupby(["movieId"], as_index = False, sort = False).sum().rename(columns = {"TF-IDF-Sq": "TF-IDF-Sq-sum"})[["movieId","TF-IDF-Sq-sum"]]
Vect_len["vect_len"] = np.sqrt(Vect_len[["TF-IDF-Sq-sum"]].sum(axis=1))
TF = pd.merge(TF,Vect_len,on = "movieId", how = "left", sort = False)
TF["TAG_WT"]=TF["TF-IDF"]/TF["vect_len"]

# calculate user profile for each user
Ratings_filter=Ratings[Ratings["rating"]>=3.5]
distinct_users=np.unique(Ratings["userId"])

#only doing one cal
user_tag_pref=pd.DataFrame()
i=1
for user in distinct_users[0:1]:
 user_data = Ratings_filter[Ratings_filter["userId"]==user]
 user_data = pd.merge(TF,user_data,on = "movieId", how = "inner", sort = False)
 user_data1 = user_data.groupby(["tag"], as_index = False, sort = False).sum().rename(columns = {"TAG_WT": "tag_pref"})[["tag","tag_pref"]]
 user_data1["user"]=user
 user_tag_pref = user_tag_pref.append(user_data1, ignore_index=True)
 i=i+1
 
#calculate the cosine similarity between the two vectors.
tag_merge_all=pd.DataFrame()

i=1
for user in distinct_users[0:1]:
 user_tag_pref_all= user_tag_pref[user_tag_pref["user"]==user]
 distinct_movies = np.unique(TF["movieId"])
 j=1
 for movie in distinct_movies:
    TF_Movie= TF[TF["movieId"]==movie]
    tag_merge = pd.merge(TF_Movie,user_tag_pref_all,on = "tag", how = "left", sort = False)
    tag_merge["tag_pref"]=tag_merge["tag_pref"].fillna(0)
    tag_merge["tag_value"]=tag_merge["TAG_WT"]*tag_merge["tag_pref"]
    TAG_WT_val=np.sqrt(np.sum(np.square(tag_merge["TAG_WT"]), axis=0))
    tag_pref_val=np.sqrt(np.sum(np.square(user_tag_pref_all["tag_pref"]), axis=0))
    tag_merge_final = tag_merge.groupby(["user","movieId"])[["tag_value"]].sum().rename(columns = {"tag_value": "Rating"}).reset_index()
 
    tag_merge_final["Rating"]=tag_merge_final["Rating"]/(TAG_WT_val*tag_pref_val)
 
    tag_merge_all = tag_merge_all.append(tag_merge_final, ignore_index=True)
    j=j+1
 i=i+1
tag_merge_all=tag_merge_all.sort_values(by=["user","Rating"]).reset_index()

File: Content Based RS/Function.R
rm(list=ls())

#Load library
library(dplyr)

#Read files
ratings <- read.csv(file.choose(),header = TRUE,stringsAsFactors = FALSE)
movies <- read.csv(file.choose(),header = TRUE,stringsAsFactors = FALSE)
tags <- read.csv(file.choose(),header = TRUE,stringsAsFactors = FALSE)

#Calculating the TF value and IDF value and multiplying together to get TF-IDF value
TF <- tags %>% group_by(movieId,tag) %>% summarise(tag_count_TF=n())
TF_unique <- TF[!duplicated(TF[c("movieId","tag")]),]
DF <- TF_unique %>% group_by(tag) %>% summarise(tag_count_DF=n())
DF$IDF <- log10(length(unique(TF_unique$movieId))) - log10(DF$tag_count_DF)
TF_unique <- merge(TF_unique,DF,by="tag",all.x = TRUE,sort=FALSE)[,c(2,1,3,4,5)]
TF_unique$TF_IDF <- TF_unique$tag_count_TF * TF_unique$IDF

#Calculating the unit length vector by dividing TF-IDF value with the vector length of a particular movie.
VectLen <- TF[,c("movieId","TF_IDF")]
VectLen$TFIDF_sq <- VectLen$TF_IDF^2
VectLenSum <- VectLen %>% group_by(movieId) %>% summarise(TFIDF_sq_sum=sum(TFIDF_sq))
VectLenSum$vect_len <- sqrt(VectLenSum$TFIDF_sq_sum)
TF_unique <- merge(TF_unique,VectLenSum,by="movieId",all.x = TRUE,sort=FALSE)
TF_unique$TAG_WT <- TF_unique$TF_IDF/TF_unique$vect_len

#Calculate user profile for each user
Ratings_filter <- ratings %>% filter(rating>=3.5)
unique_users <- unique(ratings$userId)

UserProfile <- function(user, ratingdata, TFdata){
  temprating <- ratingdata %>% filter(userId==user) %>% inner_join(TFdata, by="movieId")%>%  
                group_by(tag) %>% summarize(tag_pref=sum(TAG_WT))
  temprating$User <- rep(user,nrow(temprating))
  return(temprating)
}

user_tag_pref <- do.call(rbind,lapply(unique_users, function(x){UserProfile(x, Ratings_filter, TF_unique)}))

#Calculate the cosine similarity between the two vectors
MovCosSim <- function(movie,TFdata,userpref){
  TF_Movie <- TFdata %>% filter(movieId==movie) %>% left_join(userpref, by="tag") %>%
    mutate_at(vars(tag_pref),funs(replace(., is.na(.), 0))) %>% mutate(tag_value = TAG_WT * tag_pref)
  TAG_WT_val <- sqrt(sum((TF_Movie$TAG_WT)^2))
  tag_pref_val <- sqrt(sum((userpref$tag_pref)^2))
  tag_merge <- TF_Movie %>% group_by(User, movieId) %>% summarize(Rating=sum(tag_value))
  tag_merge$Rating <- tag_merge$Rating/(TAG_WT_val*tag_pref_val)
  return(tag_merge)
}

UserCosSim <- function(user,userdata,TFdata){
  user_tag_pref_all <- userdata %>% filter(User==user)
  unique_movies <- unique(TF$movieId)
  tag_merge <- do.call(rbind, lapply(unique_movies, function(x){MovCosSim(x,TFdata,user_tag_pref_all)}))
  return(tag_merge)
}

TagMerge <- do.call(rbind,lapply(unique_users, function(x){UserCosSim(x,user_tag_pref,TF_unique)}))
TagMerge <- TagMerge %>% na.omit() %>% arrange(User, Rating)
